cpp/lib/ResultSetArrow.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2021 Snowflake Computing, Inc. All rights reserved.
+ */
+
+#include <cmath>
+#include <iomanip>
+#include <memory>
+#include <sstream>
+#include <time.h>
+
+#include "../logger/SFLogger.hpp"
+#include "ResultSetArrow.hpp"
+#include "result_set_arrow.h"
+#include "results.h"
+
+
+namespace Snowflake
+{
+namespace Client
+{
+
+
+ResultSetArrow::ResultSetArrow() :
+    Snowflake::Client::ResultSet()
+{
+    m_queryResultFormat = QueryResultFormat::ARROW;
+    m_bufferBuilder = std::make_shared<arrow::BufferBuilder>();
+}
+
+ResultSetArrow::ResultSetArrow(
+    cJSON * initialChunk,
+    SF_COLUMN_DESC * metadata,
+    std::string tzString
+) :
+    ResultSet(metadata, tzString)
+{
+    m_queryResultFormat = QueryResultFormat::ARROW;
+    m_bufferBuilder = std::make_shared<arrow::BufferBuilder>();
+    init(initialChunk);
+}
+
+// Public methods ==================================================================================
+
+SF_STATUS STDCALL ResultSetArrow::appendChunk(cJSON * chunk)
+{
+    // Decode Base64-encoded Arrow-format rowset of the chunk.
+    const char * base64Rowset = snowflake_cJSON_GetStringValue(chunk);
+    if (base64Rowset == NULL)
+    {
+        CXX_LOG_ERROR("Unable to retrieve value of key 'base64Rowset'.");
+        return SF_STATUS_ERROR_BAD_RESPONSE;
+    }
```

Comment:
And some times the server sends empty base64Rowset in response and a subsequent chunk may have the result.